File: April 2020/20 - Construct BST from Preorder Traversal/Solution.py
class Solution:
    def bstFromPreorder(self, preorder: List[int]) -> TreeNode:
        if not preorder:
            return None;
        
        node = TreeNode(preorder[0])
        i = 1
        while i < len(preorder) and preorder[i] < node.val:
            i += 1
            
        node.left = self.bstFromPreorder(preorder[1:i])
        node.right = self.bstFromPreorder(preorder[i:])
        
        return node

File: April 2020/20 - Construct BST from Preorder Traversal/Solution.cpp
class Solution {
public:
    TreeNode *bstFromPreorder(vector<int> &preorder, int parent_value = INT_MAX) {
        return Helper(preorder, 0, preorder.size() - 1);
    }
private:
    TreeNode *Helper(vector<int> &preorder, int start, int end) {
        if(start > end)
            return NULL;
        
        TreeNode *node = new TreeNode(preorder[start]);
        int i = start;
        while(i <= end) {
            if(preorder[i] > node->val)
                break;
            ++i;
        }
        
        node->left = Helper(preorder, start + 1, i - 1);
        node->right = Helper(preorder, i, end);
        return node;
    }
};